jgrapht-core/src/main/java/org/jgrapht/alg/shortestpath/YenShortestPathIterator.java

```diff
@@ -0,0 +1,506 @@
+     * First masks all edges and vertices of the provided {@code path} except for the {@link #sink}.
+     * Then for each path in the {@link #resultList} that coincides in the {@code path}
+     * until the {@code pathDeviation} masks the edge between the {@code pathDeviation} and its
+     * successor in this path.
+     *
+     * @param path               path to mask vertices and edges of
+     * @param pathDeviation      deviation vertex of the path
+     * @param pathDeviationIndex index of the deviation vertex in the
+     *                           vertices list of the path
+     * @return pair of sets of masked vertices and edges
+     */
+    private Pair<Set<V>, Set<E>> getMaskedVerticesAndEdges(GraphPath<V, E> path,
+                                                           V pathDeviation, int pathDeviationIndex) {
+        List<V> pathVertices = path.getVertexList();
+        Set<V> maskedVertices = new HashSet<>();
+        Set<E> maskedEdges = new HashSet<>();
+
+        int pathVerticesSize = pathVertices.size();
+
+        // mask vertices and edges of the current path
+        for (int i = 0; i < pathVerticesSize - 1; i++) {
+            V v = pathVertices.get(i);
+            maskedVertices.add(v);
+            maskedEdges.add(graph.getEdge(v, pathVertices.get(i + 1)));
+        }
+
+        // mask corresponding edges of coinciding paths
+        int resultListSize = resultList.size();
+        for (int i = 0; i < resultListSize - 1; i++) { // the vertex of the current paths has already been masked
+            GraphPath<V, E> resultPath = resultList.get(i);
+            List<V> resultPathVertices = resultPath.getVertexList();
+            int deviationIndex = resultPathVertices.indexOf(pathDeviation);
+
+            if (deviationIndex < 0 || deviationIndex != pathDeviationIndex ||
+                    !equalLists(pathVertices, resultPathVertices, deviationIndex)) {
+                continue;
+            }
+
+            V successor = resultPathVertices.get(deviationIndex + 1);
+            maskedEdges.add(graph.getEdge(pathDeviation, successor));
+        }
+        return Pair.of(maskedVertices, maskedEdges);
+    }
+
+    /**
+     * Builds a candidate path based on the information provided in the
+     * methods parameters. First adds the root part of the candidate by
+     * traversing the vertices and edges of the {@code path} until the
+     * {@code recoverVertexIndex}. Then adds vertices and edges of the
+     * {@code spurPath}.
+     *
+     * @param path               path the candidate path deviates from
+     * @param recoverVertexIndex vertex that is being recovered
+     * @param spurPath           spur path of the candidate
+     * @return candidate path
+     */
+    private GraphPath<V, E> getCandidatePath(GraphPath<V, E> path, int recoverVertexIndex, GraphPath<V, E> spurPath) {
+        List<V> pathVertices = path.getVertexList();
+
+        LinkedList<V> candidatePathVertices = new LinkedList<>();
```

Comment:
Is there any reason to use LinkedList instead of ArrayList?
Also, use List interface as the type of the variable unless the actual type is needed